Adding minimal API to manage backups

cosimo opened the pull request:
When enabling the `backup` service in a couchbase cluster, it's then possible to use the [Couchbase Backup Service REST API](https://docs.couchbase.com/server/current/rest-api/backup-rest-api.html) to setup backup plans and repositories.
This PR adds minimal functionality to couchbase-cluster-admin to interact with the Backup Service Rest API.

cosimo-kahoot committed:
Small refactoring introducing an exceptions.py module

sveniu commented on couchbase_cluster_admin/cluster.py:
```diff
@@ -384,7 +386,7 @@ def create_user(self, username, user_config: dict):
         )
         if resp.status_code != 200:
-            raise Exception(f"Failed to create user: {resp.text}")
+            raise UserCreationException(resp)
```
```suggestion
            raise UserCreationException(resp.text)
```

cosimo replied:
Fixed in [f53ab3b](https://github.com/sveniu/couchbase-cluster-admin/pull/7/commits/f53ab3b180ac3ff7cf8681ecc7b0f4c94416a48f)

sveniu reviewed (commented):
Nice! Minor typo only; suggested a change.

cosimo-kahoot committed:
Fixup: UserCreationException takes resp.text

cosimo-kahoot committed:
Tentative backup plan/repository creation APIs

Not working yet, for unknown reasons. The error I'm getting is:

```
couchbase_cluster_admin.exceptions.BackupPlanCreationException: {"status":400,"msg":"invalid request body","extras":"invalid character 'd' looking for beginning of value"}
```

cosimo-kahoot committed:
Fixed incorrect logging of the ReadTimeout exception

Also added the missing import for the `logging` module.

cosimo-kahoot committed:
Whitespace changes only. Ignore.

cosimo-kahoot committed:
Use JSON payloads to send backup service API requests

We used form-encoded values everywhere, but the backup service API requires
JSON payloads, hence encode the settings in JSON before sending a request.

cosimo-kahoot committed:
Added API calls to set/delete alternate node addresses

Initially thought this would be needed for a correct backup/restore run, but
turned out it was not the problem after all.

cosimo-kahoot committed:
Added set_gsi_settings() method

Before one can restore indexes from a backup, one needs to set the default
index storage mode, and this is how.

Docs:
https://docs.couchbase.com/server/current/rest-api/post-settings-indexes.html

cosimo-kahoot committed:
Removed the automatic update of the address when renaming a node

When referring to nodes from an external network (f.ex. outside amazon aws network)
and attempting to rename a node to the internal IP address, one would of course
lose visibility, so make the address update optional after a rename

sveniu commented on couchbase_cluster_admin/cluster.py:
```diff
@@ -1,7 +1,27 @@
+    SetAlternateAddressException,
+    SetAuthenticationException,
+    SetClusterNameException,
+    SetGsiSettingsExceptin,
```
```suggestion
    SetGsiSettingsException,
```

cosimo replied:
Thanks! Fixed in 8acf35c34a979ab1f582e04ba3ce7d0c91a62352

sveniu commented on couchbase_cluster_admin/cluster.py:
```diff
@@ -425,3 +436,115 @@ def set_autofailover(self, settings: dict):
+            data=gsi_settings,
+        )
+        if resp.status_code != 200:
+            raise SetGsiSettingsExceptin(resp.text)
```
```suggestion
            raise SetGsiSettingsException(resp.text)
```

cosimo replied:
Fixed in 8acf35c34a979ab1f582e04ba3ce7d0c91a62352

sveniu commented on couchbase_cluster_admin/exceptions.py:
```diff
@@ -0,0 +1,62 @@
+    pass
+
+
+class SetGsiSettingsExceptin(Exception):
```
```suggestion
class SetGsiSettingsException(Exception):
```

cosimo replied:
Fixed in 8acf35c34a979ab1f582e04ba3ce7d0c91a62352

sveniu commented on couchbase_cluster_admin/cluster.py:
```diff
@@ -1,7 +1,27 @@
 import time
 
 from .client import BaseClient
+from .exceptions import (
```
`import *` or would you want to keep it explicit?

cosimo replied:
Used `import *` in 4f80480a

sveniu commented on couchbase_cluster_admin/cluster.py:
```diff
@@ -425,3 +436,115 @@ def set_autofailover(self, settings: dict):
+            method="POST",
+            # Backup plan settings is a multiple-level nested data structure.
+            # Requests will mangle it if we don't explicitly encode it to JSON.
+            data=json.dumps(plan_settings),
```
I'm not sure, but I think this will end up sending JSON-encoded data with Content-Type:application/x-www-form-urlencoded, and I assume Couchbase just accepts that without issue. I'm fine with merging this as is, as it clearly works.

There are two alternatives though: Modify the content-type header, or use the [json=](https://requests.readthedocs.io/en/latest/user/quickstart/#more-complicated-post-requests) param instead of data=.

cosimo replied:
Done now in b657120

sveniu reviewed (commented):
LGTM, with minor comments.

cosimo-kahoot committed:
Review: fixed typo in SetGsiSettingsException class name

cosimo-kahoot committed:
Review: moved from explicit exception imports to `import *`

cosimo-kahoot committed:
Review: use requests `json` instead of explicit json encoding

cosimo-kahoot committed:
Fixed wrong type for set_node_alternate_address() param

cosimo-kahoot committed:
Added unit tests for new set methods

* `set_node_alternate_address()`
* `delete_node_alternate_address()`
* `set_gsi_settings()`

cosimo-kahoot committed:
Return current GSI settings after setting

This is useful in the integration tests, to verify that what we've set is
coming back to us correctly.

cosimo-kahoot committed:
Added return value to the `set_gsi_settings()` api call

It's in fact possible to use `resp.json()` as Couchbase sets the content-type
correctly. We weren't doing it in our synthetic unit tests, but we're doing it
now. We use this now in the integration tests to verify that our gsi settings
have been really modified.

cosimo-kahoot committed:
Bumped version to 0.2.0

cosimo-kahoot committed:
Added support for requests with JSON request body

cosimo-kahoot committed:
Added unit tests for the backup service API

sveniu reviewed (approved):
This looks great! I'll merge shortly.
